<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Zachary McCaw" />


<title>Rank Normal Omnibus Association Test</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Rank Normal Omnibus Association Test</h1>
<h4 class="author"><em>Zachary McCaw</em></h4>
<h4 class="date"><em>Updated: 10/17/18</em></h4>



<div id="contents" class="section level2">
<h2>Contents</h2>
<ul>
<li><a href="#example-data">Example Data</a></li>
<li><a href="#genetic-association-testing">Genetic Association Testing</a></li>
<li><a href="#run-time">Run Time</a></li>
<li><a href="#missingness">Missingness</a></li>
</ul>
</div>
<div id="example-data" class="section level2">
<h2>Example Data</h2>
<p>Simulated data is available for <span class="math inline">\(10^{3}\)</span> subjects. Covariates include <code>Age</code> and <code>Sex</code>. Structure adjustments include the first two principal components, <code>pc1</code> and <code>pc2</code>, of the centered and scaled subject by locus genotype matrix. Genotypes at <span class="math inline">\(10^{2}\)</span> loci on chromosome one are included. Sample minor allele frequency for each locus falls in the range <span class="math inline">\([0.110, 0.360]\)</span>. Two independent phenotypes are available. <code>YN</code> has normally distributed residuals, while the log of <code>YL</code> has normally distributed residuals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RNOmni);
## Example data
X =<span class="st"> </span>RNOmni::X;
<span class="kw">cat</span>(<span class="st">&quot;Covariates</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">round</span>(<span class="kw">head</span>(X),<span class="dt">digits=</span><span class="dv">2</span>);
<span class="kw">cat</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">cat</span>(<span class="st">&quot;Structure Adjustments</span><span class="ch">\n</span><span class="st">&quot;</span>);
S =<span class="st"> </span>RNOmni::S;
<span class="kw">round</span>(<span class="kw">head</span>(S),<span class="dt">digits=</span><span class="dv">2</span>);
<span class="kw">cat</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">cat</span>(<span class="st">&quot;Genotype Matrix</span><span class="ch">\n</span><span class="st">&quot;</span>);
G =<span class="st"> </span>RNOmni::Geno;
G[<span class="dv">1</span>:<span class="dv">6</span>,<span class="dv">1</span>:<span class="dv">6</span>];
<span class="kw">cat</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">cat</span>(<span class="st">&quot;Sample Minor Allele Frequency</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">summary</span>(<span class="kw">apply</span>(G,<span class="dt">MARGIN=</span><span class="dv">2</span>,<span class="dt">FUN=</span>mean)/<span class="dv">2</span>);
<span class="kw">cat</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">cat</span>(<span class="st">&quot;Phenotypes</span><span class="ch">\n</span><span class="st">&quot;</span>);
Y =<span class="st"> </span>RNOmni::Y;
<span class="kw">round</span>(<span class="kw">head</span>(Y),<span class="dt">digits=</span><span class="dv">2</span>);</code></pre></div>
<pre><code>## Covariates
##        Age Sex
## [1,] 50.75   0
## [2,] 46.92   1
## [3,] 52.81   0
## [4,] 48.42   0
## [5,] 48.05   1
## [6,] 47.75   1
## 
## Structure Adjustments
##        pc1    pc2
## [1,] -1.90 -11.54
## [2,]  5.43   1.75
## [3,] 16.76   3.23
## [4,] 19.29 -13.15
## [5,]  2.92  -1.49
## [6,] 10.92   7.55
## 
## Genotype Matrix
##      s1 s2 s3 s4 s5 s6
## [1,]  0  0  0  1  0  0
## [2,]  1  0  2  0  1  0
## [3,]  0  0  0  1  1  0
## [4,]  0  0  2  2  1  1
## [5,]  2  1  0  2  1  1
## [6,]  0  0  0  0  0  0
## 
## Sample Minor Allele Frequency
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##   0.110   0.185   0.215   0.219   0.250   0.360 
## 
## Phenotypes
##        YN    YL
## [1,] 2.42 27.29
## [2,] 4.15 42.35
## [3,] 2.65 64.34
## [4,] 3.13  7.97
## [5,] 5.08 97.49
## [6,] 2.14 17.84</code></pre>
</div>
<div id="genetic-association-testing" class="section level2">
<h2>Genetic Association Testing</h2>
<div id="basic-association-test" class="section level4">
<h4>Basic Association Test</h4>
<p><code>BAT</code> regresses the untransformed phenotype <span class="math inline">\(y\)</span> on genotype <span class="math inline">\(g\)</span>, covariates <span class="math inline">\(X\)</span>, and adjustments for substructure <span class="math inline">\(S\)</span>. The Wald statistic is used to calculate a <span class="math inline">\(p\)</span>-value assessing the null hypothesis of no genotypic effect. The output is a numeric vector, with one <span class="math inline">\(p\)</span>-value per locus in <span class="math inline">\(G\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Basic Association Test</span>
p1.bat =<span class="st"> </span>RNOmni::<span class="kw">BAT</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S);
<span class="kw">round</span>(<span class="kw">head</span>(p1.bat),<span class="dt">digits=</span><span class="dv">3</span>);</code></pre></div>
<pre><code>## [1] 0.654 0.407 0.770 0.555 0.877 0.722</code></pre>
</div>
<div id="direct-inverse-normal-transformation" class="section level4">
<h4>Direct Inverse Normal Transformation</h4>
<p><code>DINT</code> directly applies the rank-based inverse normal transformation (INT) to the phenotype <span class="math inline">\(y\)</span>. The transformed phenotype <span class="math inline">\(\text{INT}(y)\)</span> is regressed on genotype <span class="math inline">\(g\)</span>, covariates <span class="math inline">\(X\)</span>, and adjustments for population structure <span class="math inline">\(S\)</span>. The Wald statistic is used to calculate a <span class="math inline">\(p\)</span>-value assessing the null hypothesis of no genotypic effect. The output is a numeric vector, with one <span class="math inline">\(p\)</span>-value per locus in <span class="math inline">\(G\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Direct INT Test</span>
p1.dint =<span class="st"> </span>RNOmni::<span class="kw">DINT</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S);
<span class="kw">round</span>(<span class="kw">head</span>(p1.dint),<span class="dt">digits=</span><span class="dv">3</span>);</code></pre></div>
<pre><code>## [1] 0.632 0.380 0.740 0.568 0.845 0.739</code></pre>
</div>
<div id="partially-indirect-inverse-normal-transformation" class="section level4">
<h4>Partially Indirect Inverse Normal Transformation</h4>
<p><code>PIINT</code> implements a two-stage, INT-based association test. In the first stage, the phenotype <span class="math inline">\(y\)</span> is regressed on covariates <span class="math inline">\(X\)</span> to obtain residuals <span class="math inline">\(e\)</span>. In the second stage, the transformed residuals <span class="math inline">\(\text{INT}(e)\)</span> are regressed on genotype <span class="math inline">\(g\)</span> and adjustments for population structure <span class="math inline">\(S\)</span>. The Wald statistic is used to calculate a <span class="math inline">\(p\)</span>-value assessing the null hypothesis of no genotypic effect. The output is a numeric vector, with one <span class="math inline">\(p\)</span>-value per locus in <span class="math inline">\(G\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Partially Indirect INT Test</span>
p1.piint =<span class="st"> </span>RNOmni::<span class="kw">PIINT</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S);
<span class="kw">round</span>(<span class="kw">head</span>(p1.piint),<span class="dt">digits=</span><span class="dv">3</span>);</code></pre></div>
<pre><code>## [1] 0.674 0.389 0.731 0.543 0.852 0.709</code></pre>
</div>
<div id="omnibus-inverse-normal-transformation-method" class="section level4">
<h4>Omnibus Inverse Normal Transformation Method</h4>
<p><code>RNOmni</code> implements an adaptive, INT-based association test. In the omnibus test, <code>DINT</code> and <code>PIINT</code> are each applied, then an omnibus statistic is calculated based on whichever approach provides more evidence against the null. Estimation of a <span class="math inline">\(p\)</span>-value for the omnibus statistic requires an estimate of the correlation <span class="math inline">\(\rho\)</span> between the test statistics provided by <code>DINT</code> and <code>PIINT</code>. When the sample size and number of loci are both relatively large, an efficient estimate of <span class="math inline">\(\rho\)</span> is obtained by averaging across loci. When the sample size or number of loci are smaller, bootstrap can be used to provide a locus-specific estimate of <span class="math inline">\(\rho\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cat</span>(<span class="st">&quot;Omnibus Test, Average Correaltion Method</span><span class="ch">\n</span><span class="st">&quot;</span>);
p1.omni.avg =<span class="st"> </span>RNOmni::<span class="kw">RNOmni</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S,<span class="dt">method=</span><span class="st">&quot;AvgCorr&quot;</span>);
<span class="kw">round</span>(<span class="kw">head</span>(p1.omni.avg),<span class="dt">digits=</span><span class="dv">3</span>);
<span class="kw">cat</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">cat</span>(<span class="st">&quot;Omnibus Test, Bootstrap Correaltion Method</span><span class="ch">\n</span><span class="st">&quot;</span>);
p1.omni.boot =<span class="st"> </span>RNOmni::<span class="kw">RNOmni</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S,<span class="dt">method=</span><span class="st">&quot;Bootstrap&quot;</span>,<span class="dt">B=</span><span class="dv">100</span>,<span class="dt">cores=</span><span class="dv">12</span>);
<span class="kw">round</span>(<span class="kw">head</span>(p1.omni.boot),<span class="dt">digits=</span><span class="dv">3</span>);
<span class="kw">cat</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);</code></pre></div>
<pre><code>## Omnibus Test, Average Correaltion Method
##       DINT PIINT RNOmni
## [1,] 0.632 0.674  0.759
## [2,] 0.380 0.389  0.508
## [3,] 0.740 0.731  0.842
## [4,] 0.568 0.543  0.677
## [5,] 0.845 0.852  0.923
## [6,] 0.739 0.709  0.824
## 
## Omnibus Test, Bootstrap Correaltion Method
##       DINT PIINT RNOmni
## [1,] 0.632 0.674  0.676
## [2,] 0.380 0.389  0.400
## [3,] 0.740 0.731  0.749
## [4,] 0.568 0.543  0.585
## [5,] 0.845 0.852  0.858
## [6,] 0.739 0.709  0.732</code></pre>
</div>
<div id="comparison-of-p-values" class="section level4">
<h4>Comparison of p-values</h4>
<p>The following table compares the estimated <span class="math inline">\(p\)</span>-values for the normal phenotype obtained by each association test. Note that for the normal phenotype, all approaches provided valid inference in simulation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">cat</span>(<span class="st">&quot;Normal Phenotype, Combined Results</span><span class="ch">\n</span><span class="st">&quot;</span>);
P =<span class="st"> </span><span class="kw">cbind</span>(<span class="st">&quot;BAT&quot;</span>=p1.bat,<span class="st">&quot;DINT&quot;</span>=p1.dint,<span class="st">&quot;PIINT&quot;</span>=p1.piint,
          <span class="st">&quot;Omni.Avg&quot;</span>=p1.omni.avg[,<span class="st">&quot;RNOmni&quot;</span>],<span class="st">&quot;Omni.Boot&quot;</span>=p1.omni.boot[,<span class="st">&quot;RNOmni&quot;</span>]);
<span class="kw">show</span>(<span class="kw">round</span>(<span class="kw">head</span>(P),<span class="dt">digits=</span><span class="dv">4</span>));
<span class="kw">cat</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">cat</span>(<span class="st">&quot;Empirical Size</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">apply</span>((P&lt;=<span class="fl">0.05</span>),<span class="dt">MARGIN=</span><span class="dv">2</span>,<span class="dt">FUN=</span>mean);</code></pre></div>
<pre><code>## Normal Phenotype, Combined Results
##         BAT   DINT  PIINT Omni.Avg Omni.Boot
## [1,] 0.6536 0.6323 0.6745   0.7595    0.6761
## [2,] 0.4070 0.3799 0.3890   0.5084    0.3995
## [3,] 0.7697 0.7400 0.7315   0.8418    0.7493
## [4,] 0.5547 0.5683 0.5427   0.6770    0.5847
## [5,] 0.8766 0.8454 0.8520   0.9234    0.8579
## [6,] 0.7219 0.7391 0.7092   0.8242    0.7325
## 
## Empirical Size
##       BAT      DINT     PIINT  Omni.Avg Omni.Boot 
##      0.02      0.02      0.02      0.02      0.02</code></pre>
<p>The following table compares the estimated <span class="math inline">\(p\)</span>-values for the log-normal phenotype obtained by each association test. Note that the basic association test did not consistently control the type I error in simulation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Estimate p-values</span>
p2.bat =<span class="st"> </span>RNOmni::<span class="kw">BAT</span>(<span class="dt">y=</span>Y[,<span class="dv">2</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S);
p2.omni.avg =<span class="st"> </span>RNOmni::<span class="kw">RNOmni</span>(<span class="dt">y=</span>Y[,<span class="dv">2</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S,<span class="dt">method=</span><span class="st">&quot;AvgCorr&quot;</span>);
p2.omni.boot =<span class="st"> </span>RNOmni::<span class="kw">RNOmni</span>(<span class="dt">y=</span>Y[,<span class="dv">2</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S,<span class="dt">method=</span><span class="st">&quot;Bootstrap&quot;</span>,<span class="dt">B=</span><span class="dv">100</span>,<span class="dt">cores=</span><span class="dv">12</span>);
<span class="kw">cat</span>(<span class="st">&quot;Log Normal Phenotype, Combined Results</span><span class="ch">\n</span><span class="st">&quot;</span>);
P2 =<span class="st"> </span><span class="kw">cbind</span>(<span class="st">&quot;BAT&quot;</span>=p2.bat,<span class="st">&quot;DINT&quot;</span>=p2.omni.avg[,<span class="st">&quot;DINT&quot;</span>],<span class="st">&quot;PIINT&quot;</span>=p2.omni.avg[,<span class="st">&quot;PIINT&quot;</span>],
          <span class="st">&quot;Omni.Avg&quot;</span>=p2.omni.avg[,<span class="st">&quot;RNOmni&quot;</span>],<span class="st">&quot;Omni.Boot&quot;</span>=p2.omni.boot[,<span class="st">&quot;RNOmni&quot;</span>]);
<span class="kw">show</span>(<span class="kw">round</span>(<span class="kw">head</span>(P2),<span class="dt">digits=</span><span class="dv">4</span>));
<span class="kw">cat</span>(<span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">cat</span>(<span class="st">&quot;Empirical Size</span><span class="ch">\n</span><span class="st">&quot;</span>);
<span class="kw">apply</span>((P2&lt;=<span class="fl">0.05</span>),<span class="dt">MARGIN=</span><span class="dv">2</span>,<span class="dt">FUN=</span>mean);</code></pre></div>
<pre><code>## Log Normal Phenotype, Combined Results
##         BAT   DINT  PIINT Omni.Avg Omni.Boot
## [1,] 0.8343 0.3242 0.8252   0.4798    0.4200
## [2,] 0.3922 0.5092 0.4808   0.6552    0.6007
## [3,] 0.6859 0.8284 0.5796   0.7504    0.6897
## [4,] 0.5644 0.6298 0.6206   0.7863    0.7460
## [5,] 0.8994 0.7919 0.8406   0.9129    0.8754
## [6,] 0.1769 0.2082 0.4896   0.3292    0.2777
## 
## Empirical Size
##       BAT      DINT     PIINT  Omni.Avg Omni.Boot 
##      0.01      0.03      0.01      0.01      0.01</code></pre>
</div>
</div>
<div id="run-time" class="section level2">
<h2>Run Time</h2>
<p>During package development, the <code>BAT</code>, <code>DINT</code>, and <code>PIINT</code> each took a median of <span class="math inline">\(\sim 25\)</span> ms to perform <span class="math inline">\(10^2\)</span> association tests for <span class="math inline">\(10^3\)</span> subjects. <code>RNOmni</code> using average correlation, which interval performs both <code>DINT</code> and <code>PIINT</code>, required a median of <span class="math inline">\(\sim 65\)</span> ms. Using bootstrap to calculate position specific correlations increased the run time of <code>RNOmni</code> by a factor of <span class="math inline">\(10\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RNOmni);
microbenchmark::<span class="kw">microbenchmark</span>(<span class="kw">BAT</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S),<span class="kw">DINT</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S),<span class="kw">PIINT</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S),
                               <span class="kw">RNOmni</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S,<span class="dt">method=</span><span class="st">&quot;AvgCorr&quot;</span>),
                               <span class="kw">RNOmni</span>(<span class="dt">y=</span>Y[,<span class="dv">1</span>],<span class="dt">G=</span>G,<span class="dt">X=</span>X,<span class="dt">S=</span>S,<span class="dt">method=</span><span class="st">&quot;Bootstrap&quot;</span>,<span class="dt">B=</span><span class="dv">100</span>,<span class="dt">cores=</span><span class="dv">12</span>));</code></pre></div>
<pre><code>## Unit: milliseconds
##                                                                                     expr
##                                                     BAT(y = Y[, 1], G = G, X = X, S = S)
##                                                    DINT(y = Y[, 1], G = G, X = X, S = S)
##                                                   PIINT(y = Y[, 1], G = G, X = X, S = S)
##                              RNOmni(y = Y[, 1], G = G, X = X, S = S, method = &quot;AvgCorr&quot;)
##  RNOmni(y = Y[, 1], G = G, X = X, S = S, method = &quot;Bootstrap&quot;,      B = 100, cores = 12)
##        min        lq      mean    median        uq      max neval
##   12.53513  13.58272  22.33492  14.04694  15.53754 174.9711   100
##   12.73102  13.53048  16.50133  14.16752  15.48039 156.9364   100
##   10.88675  11.72506  19.35793  12.59122  15.09683 172.4945   100
##   39.85082  41.92245  52.14607  43.66620  49.87676 201.3037   100
##  400.11306 421.55207 527.18372 575.44465 587.81182 604.1654   100</code></pre>
</div>
<div id="missingness" class="section level2">
<h2>Missingness</h2>
<p>Observations missing either the phenotype <span class="math inline">\(y\)</span> or the the substructure adjustments <span class="math inline">\(S\)</span> are excluded. Missing covariates <span class="math inline">\(X\)</span> are imputed to the median of the observed values. An observation missing genotype information <span class="math inline">\(G\)</span> is excluded from association testing only at those loci were genotype is unobserved.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Introduce Missingness</span>
y.m =<span class="st"> </span>Y[,<span class="dv">1</span>];
y.m[<span class="kw">sample</span>(<span class="kw">length</span>(y.m),<span class="dt">size=</span><span class="dv">10</span>,<span class="dt">replace=</span>F)] =<span class="st"> </span><span class="ot">NA</span>;
G.m =<span class="st"> </span>G;
G.m[<span class="kw">sample</span>(<span class="kw">length</span>(G.m),<span class="dt">size=</span><span class="dv">1000</span>,<span class="dt">replace=</span>F)] =<span class="st"> </span><span class="ot">NA</span>;
X.m =<span class="st"> </span>X;
X.m[<span class="kw">sample</span>(<span class="kw">length</span>(X.m),<span class="dt">size=</span><span class="dv">100</span>,<span class="dt">replace=</span>F)] =<span class="st"> </span><span class="ot">NA</span>;
S.m =<span class="st"> </span>S;
S.m[<span class="kw">sample</span>(<span class="kw">length</span>(S.m),<span class="dt">size=</span><span class="dv">10</span>,<span class="dt">replace=</span>F)] =<span class="st"> </span><span class="ot">NA</span>;
<span class="co"># Association Testing after Missingness</span>
p1.bat.m =<span class="st"> </span>RNOmni::<span class="kw">BAT</span>(<span class="dt">y=</span>y.m,<span class="dt">G=</span>G.m,<span class="dt">X=</span>X.m,<span class="dt">S=</span>S.m);
p1.dint.m =<span class="st"> </span>RNOmni::<span class="kw">DINT</span>(<span class="dt">y=</span>y.m,<span class="dt">G=</span>G.m,<span class="dt">X=</span>X.m,<span class="dt">S=</span>S.m);
p1.piint.m =<span class="st"> </span>RNOmni::<span class="kw">PIINT</span>(<span class="dt">y=</span>y.m,<span class="dt">G=</span>G.m,<span class="dt">X=</span>X.m,<span class="dt">S=</span>S.m);
p1.omni.avg.m =<span class="st"> </span>RNOmni::<span class="kw">RNOmni</span>(<span class="dt">y=</span>y.m,<span class="dt">G=</span>G.m,<span class="dt">X=</span>X.m,<span class="dt">S=</span>S.m);
p1.omni.boot.m =<span class="st"> </span>RNOmni::<span class="kw">RNOmni</span>(<span class="dt">y=</span>y.m,<span class="dt">G=</span>G.m,<span class="dt">X=</span>X.m,<span class="dt">S=</span>S.m,<span class="dt">method=</span><span class="st">&quot;Bootstrap&quot;</span>,<span class="dt">B=</span><span class="dv">100</span>);</code></pre></div>
<p>Tabulation of <span class="math inline">\(p\)</span>-values in the presence of missingness:</p>
<pre><code>## Normal Phenotype in the Absence of Missingness, Combined Results
##         BAT   DINT  PIINT Omni.Avg Omni.Boot
## [1,] 0.6536 0.6323 0.6745   0.7595    0.6761
## [2,] 0.4070 0.3799 0.3890   0.5084    0.3995
## [3,] 0.7697 0.7400 0.7315   0.8418    0.7493
## [4,] 0.5547 0.5683 0.5427   0.6770    0.5847
## [5,] 0.8766 0.8454 0.8520   0.9234    0.8579
## [6,] 0.7219 0.7391 0.7092   0.8242    0.7325
## 
## Normal Phenotype in the Presence of Missingness, Combined Results
##         BAT   DINT  PIINT Omni.Avg Omni.Boot
## [1,] 0.8007 0.7776 0.8432   0.8460    0.8026
## [2,] 0.4564 0.4317 0.4385   0.5226    0.4549
## [3,] 0.8130 0.7840 0.7742   0.8431    0.7942
## [4,] 0.7602 0.7808 0.7401   0.8147    0.7650
## [5,] 0.9620 0.9973 0.9722   0.9862    0.9781
## [6,] 0.7568 0.7749 0.7366   0.8117    0.7528</code></pre>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
